### PR TITLE
PR3. Add Pathofact2-style combined report generation

### PR DESCRIPTION
**Pull request by Ales-ibt**
This is PR3 out of 4.

PR3 adds the combined annotation report that merges PathoFact2, AMR, BGC, mobilome, and InterProScan results into a single per-protein output.

What's included:
 - bin/pathofact2_report.py — core Python logic: parses each tool's output and joins them on protein ID
 - modules/local/combinereporter.nf — Nextflow wrapper that calls the script
 - modules/local/reformat_amrfinder_tsv.nf — reformats AMRFinderPlus TSV to match expected column layout before report merging. This is necessary when a precomputed result of AMRfinder plus generated by the genomes catalogue pipeline or mettanotator is provided.
 - tests/pathofact_samplesheet.csv — test samplesheet for the pathofact integration test

Note: COMBINEREPORTER is not yet invoked from the main workflow — the wiring happens in PR4. This PR focuses on the script logic for review.

Thanks for reviewing!

### CODE REVIEW

File: bin/pathofact2_report.py [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+    -------
+    bgc_data
+        protein_id -> (bgc_tools, bgc_type)
+    """
+    logging.info("Parsing BGC GFF: %s", gff_file)
+
+    bgc_tools_map: DefaultDict[str, list[str]] = defaultdict(list)
+    bgc_type_map: DefaultDict[str, list[str]] = defaultdict(list)
+
+    with fileinput.hook_compressed(gff_file, "r", encoding="utf-8", errors="ignore") as handle:
```
Comment (mberacochea) [inline bin/pathofact2_report.py]: Can't we use the toolkit GFF parser?

File: bin/pathofact2_report.py [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+            bgc_tools_values = split_csv_value(attributes.get("bgc_tools", "-"))
+            bgc_tools_map[protein_id] = merge_unique_preserving_order(
+                bgc_tools_map[protein_id],
+                bgc_tools_values,
+            )
+
+            bgc_type_values: list[str] = []
+            bgc_type_values.extend(split_csv_value(attributes.get("nearest_MiBIG_class", "-")))
+            bgc_type_values.extend(split_csv_value(attributes.get("antismash_product", "-")))
+            bgc_type_values.extend(split_csv_value(attributes.get("gecco_bgc_type", "-")))
```
Comment (mberacochea) [inline bin/pathofact2_report.py]: These kind of things should be "normalized" across pipelines

**Review by mberacochea — CHANGES_REQUESTED** [review]
I worry about GFF parsing and naming rules and stuff. I know the toolkit is a bit of pain in the neck, but having things to be consistent across pipelines pays in the long run.

Should we have this GFF parsing stuff in the toolkit? Ideas?

File: bin/pathofact2_report.py (hunk 2 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py]: yes, this implementation consider the attributes we are keeping on mettanotator, CGP and ASA. I made an effort to align this output with the rules set on those pipelines

File: modules/local/combinereporter.nf [hunk 3]
```diff
@@ -0,0 +1,42 @@
+process COMBINEREPORTER {
+    tag "$meta.id"
+    label 'process_single'
+
+    conda "${moduleDir}/environment.yml"
```
Comment (timrozday-mgnify) [inline modules/local/combinereporter.nf]: This file doesn't exist

File: tests/pathofact_samplesheet.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+sample,assembly,proteins_gff,proteins_faa,virify_gff,interproscan_tsv
+MGYG000528118_sub,/hps/nobackup/rdf/metagenomics/service-team/projects/infectome/map_pathofact_dev/mobilome-annotation-pipeline/tests/data/MGYG000528118_sub/MGYG000528118_sub.fna,/hps/nobackup/rdf/metagenomics/service-team/projects/infectome/map_pathofact_dev/mobilome-annotation-pipeline/tests/data/MGYG000528118_sub/MGYG000528118_sub.gff,/hps/nobackup/rdf/metagenomics/service-team/projects/infectome/map_pathofact_dev/mobilome-annotation-pipeline/tests/data/MGYG000528118_sub/MGYG000528118_sub.faa,,
```
Comment (timrozday-mgnify) [inline tests/pathofact_samplesheet.csv]: These paths are Ales-specific, they need to be relative in the repo or use `${projectDir}`

File: bin/pathofact2_report.py [hunk 5]
```diff
@@ -0,0 +1,628 @@
+#!/usr/bin/env python3
+
+"""
+Integrate PathoFact2, AMR, mobilome, BGC, and InterProScan annotations.
```
Comment (ochkalova) [inline bin/pathofact2_report.py]: What is AMR here? If it's AMRFinder tool, it worth replace with a full name across the docstrings for clarity

File: bin/pathofact2_report.py [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+            seen.add(value)
+
+    return merged
+
+
+def join_or_dash(values: list[str]) -> str:
+    return ",".join(values) if values else "-"
+
+
+def parse_pathofact2_gff(
```
Comment (timrozday-mgnify) [inline bin/pathofact2_report.py:165]: Can each of the `parse_...` functions be refactored to use some shared code? and furthermore use a GFF library (as Martin suggested)?

File: bin/pathofact2_report.py [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+            if overlap_fraction >= min_fraction:
+                assigned_mges = merge_unique_preserving_order(assigned_mges, [mge_type])
+
+        protein_mge_map[protein_id] = join_or_dash(assigned_mges)
+
+    logging.info("Assigned MGE context to %s proteins", len(protein_mge_map))
+    return protein_mge_map
+
+
+def build_rows(
```
Comment (timrozday-mgnify) [inline bin/pathofact2_report.py:401]: This is a nice clear bit of code for building a combined output 👍

**Review by timrozday-mgnify — CHANGES_REQUESTED** [review]
Thanks @Ales-ibt , there's some nice structure in this code.

However, it's hard to review a module with this much code without at least a module test. Ideally there would be some unit tests for the Python script as well e.g. for each of the `parse_...` functions checking the typical case and some edge cases. People don't like these tests in the nextflow bin directory so perhaps `mgnify-pipelines-toolkit` is the right place for it.

in `bin/pathofact2_report.py` there is a lot of repetition with the `parse_...` functions, can they be refactored to use some shared code? or better yet a GFF library (as Martin suggested).

File: bin/pathofact2_report.py [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+        return True
+
+    path = Path(path_str)
+    if not path.exists():
+        return True
+
+    if path.stat().st_size == 0:
+        return True
+
+    return False
```
Comment (ochkalova) [inline bin/pathofact2_report.py]: This can be shortened:
```suggestion
def path_is_missing_or_empty(path_str: str) -> bool:
    if not path_str:
        return True

    path = Path(path_str)
    return not path.exists() or path.stat().st_size == 0
```

**Comment by timrozday-mgnify** [comment]
> I worry about GFF parsing and naming rules and stuff. I know the toolkit is a bit of pain in the neck, but having things to be consistent across pipelines pays in the long run.
> 
> Should we have this GFF parsing stuff in the toolkit? Ideas?

I agree, unit tests are important to check proper parsing of each of the different inputs, and it sounds like there's likely to be re-use.

File: bin/pathofact2_report.py [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+
+def split_csv_value(value: str) -> list[str]:
+    if value in {"", "-"}:
+        return []
+    return [item.strip() for item in value.split(",") if item.strip()]
+
+
+def merge_unique_preserving_order(existing: list[str], new_values: list[str]) -> list[str]:
+    seen = set(existing)
+    merged = list(existing)
```
Comment (ochkalova) [inline bin/pathofact2_report.py]: You can use `.copy()` instead of listing the list 😉 
```suggestion
    merged = existing.copy()
```

File: bin/pathofact2_report.py (hunk 6 above)
Comment (ochkalova) [reply bin/pathofact2_report.py:165]: +1 There is some repetitive code, would be good to make it more DRY

File: modules/local/reformat_amrfinder_tsv.nf [hunk 10]
```diff
@@ -0,0 +1,38 @@
+// Reformat an AMRFinderPlus TSV produced by the genomes-catalogue-pipeline so that
+// its columns align with what amr_integrator expects.
+//
+// Catalogue column layout:  protein_id=1, drug_class=11, identity=17
+// amr_integrator expected:  protein_id=1, drug_class=7,  identity=13
```
Comment (ochkalova) [inline modules/local/reformat_amrfinder_tsv.nf:5]: They are identical, aren't they? Is it the intended example? 🤔

File: modules/local/combinereporter.nf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/python:3.12.12':
+        'biocontainers/python:3.12.12'}"
+
+    input:
+    tuple val(meta), path(mobilome_gff), path(pathofact_gff), path(arg_gff), path(bgcs_gff), path(ips_tsv)
```
Comment (ochkalova) [inline modules/local/combinereporter.nf:10]: What is arg_gff? 🤔

File: bin/pathofact2_report.py [hunk 12]
```diff
@@ -0,0 +1,628 @@
...
+    if pathofact_missing and amr_missing:
+        logging.info(
+            "PathoFact2 and AMR files are both absent or empty; no seed proteins found. "
+            "Skipping downstream parsing and not generating an output file."
+        )
+        return
+
+    prots_coords: dict[str, tuple[str, int, int]] = {}
+    pathofact_data: dict[str, tuple[str, str, str, str, str]] = {}
+    amr_data: dict[str, tuple[str, str, str]] = {}
```
Comment (ochkalova) [inline bin/pathofact2_report.py:484]: In the future, you could consider using data classes instead of tuples. Those provide some cons in terms of implicity and readability. Something like:
```
from dataclasses import dataclass

@dataclass
class PathoFactData:
    vfdb_hit: str
    blastp_eval: str
    tox_prob: str
    vf_prob: str
    cdd_annotation: str

@dataclass
class AMRData:
    drug_class: str
    tool: str
    identity: str
```

File: bin/pathofact2_report.py [hunk 13]
```diff
@@ -0,0 +1,628 @@
...
+    parser.add_argument(
+        "--amr",
+        required=False,
+        default="",
+        help="Optional AMR GFF file",
+    )
+    parser.add_argument(
+        "--mobilome",
+        required=False,
+        default="",
```
Comment (ochkalova) [inline bin/pathofact2_report.py]: Just in case, you do not need to redefine the default to the empty string, by default `default` is `None` and it would work fine for your code. You can also specify `type=Path` to make automatic casting of provided paths to Path. 
This way your function `path_is_missing_or_empty` could look like:
```
def path_is_missing_or_empty(path: None | Path) -> bool:
    if not path:
        return True

    return not path.exists() or path.stat().st_size == 0
```

**Review by ochkalova — COMMENTED** [review]
Ales, great job! Thank you very much for not piling up all changes to one giant PR 💀 
It's so much easier to review. 

Logic looks good, I only left minor comments about code readability. Would be good to refactor GFFs parsing.

File: modules/local/reformat_amrfinder_tsv.nf (hunk 10 above)
Comment (Ales-ibt) [reply modules/local/reformat_amrfinder_tsv.nf:5]: It is showing the column number in wich the header is placed in each version of the output. The drug_class and identity are in different columns, that's why we need to reformat when using a pre-computed result from the genomes catalogue pipeline

File: modules/local/combinereporter.nf (hunk 11 above)
Comment (Ales-ibt) [reply modules/local/combinereporter.nf:10]: This is the GFF generated for the antimicrobial resistance genes subworkflow

File: modules/local/combinereporter.nf (hunk 3 above)
Comment (Ales-ibt) [reply modules/local/combinereporter.nf]: Good catch :D 
I have removed that line

File: bin/pathofact2_report.py (hunk 1 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py]: Yes, I have bring the toolkit to the party here

File: tests/pathofact_samplesheet.csv (hunk 4 above)
Comment (Ales-ibt) [reply tests/pathofact_samplesheet.csv]: Good catch. I removed those paths

File: bin/pathofact2_report.py (hunk 6 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py:165]: Yes, I had removed the redundant code and use the mgnify toolkit instead

File: bin/pathofact2_report.py (hunk 7 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py:401]: Thank you :D

File: bin/pathofact2_report.py (hunk 5 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py]: It is antimicrobial resistance genes annotation. Results comes from the subworkflow that runs three different tools. I had replaced AMR and BGC by the full names of the annotation type

File: bin/pathofact2_report.py (hunk 13 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py]: Change added

File: bin/pathofact2_report.py (hunk 12 above)
Comment (Ales-ibt) [reply bin/pathofact2_report.py:484]: Interesting! I will keep the code as it is by now as it is working and this is more like a coding-style suggestion :D

**Comment by Ales-ibt** [comment]
Thank you for your review. I am now using the gff io from the toolkit and I have added a test to the bin/pathofact2_report.py script on the tests/unit directory. I will tag you again just to have at least one approve before merging.

File: bin/pathofact2_report.py [hunk 14]
```diff
@@ -0,0 +1,541 @@
...
+Rules
+-----
+- Seed proteins come from PathoFact2 and/or AMR GFF files.
+- If both PathoFact2 and AMR inputs are absent/empty, the script exits
+  without generating any output.
+- BGC annotations are parsed only for proteins in the seed protein set.
+- MGE assignment is based on >= 90% CDS overlap with a mobilome feature
+  in the same contig.
+- InterProScan is optional and only SignalP entries are retained.
+- Missing values are reported as '-'.
```
Comment (mberacochea) [inline bin/pathofact2_report.py:29]: This would be useful in the docs too

Comment (Ales-ibt) [reply bin/pathofact2_report.py:29]: I will add this on PR4 where the changes in the readme includes pathofact2 outputs

**Comment by Ales-ibt** [comment]
I have assessing all the comments and changes suggested by reviewers.  I am merging here as I need to move on. Thanks for your reviews :D
